src/index/blockfilterindex.cpp

```diff
@@ -387,13 +389,30 @@ bool BlockFilterIndex::LookupFilter(const CBlockIndex* block_index, BlockFilter&
     return ReadFilterFromDisk(entry.pos, filter_out);
 }
 
-bool BlockFilterIndex::LookupFilterHeader(const CBlockIndex* block_index, uint256& header_out) const
+bool BlockFilterIndex::LookupFilterHeader(const CBlockIndex* block_index, uint256& header_out)
 {
+    {
+        // Check the headers cache
+        LOCK(m_cs_headers_cache);
+        auto header = m_headers_cache.find(block_index->GetBlockHash());
+        if (header != m_headers_cache.end()) {
+            header_out = header->second;
+            return true;
+        }
+    }
+
     DBVal entry;
     if (!LookupOne(*m_db, block_index, entry)) {
         return false;
     }
 
+    if (block_index->nHeight % CFCHECKPT_INTERVAL == 0) {
+        // Add to the headers cache
+        LOCK(m_cs_headers_cache);
+        if (m_headers_cache.size() < CF_HEADERS_CACHE_MAX_SZ) {
+            m_headers_cache.emplace(block_index->GetBlockHash(), entry.header);
```

Comment:
The test suite explicitly checks that filters for stale blocks may still be fetched.

https://github.com/jnewbery/bitcoin/blob/2020-05-cfcheckpts-cache/test/functional/p2p_blockfilters.py#L89-L101

Reply:
> we could end up with headers of orphaned blocks in the cache

Correct. If there's a re-org across checkpoint block, then the disconnect block's filter header stays in the cache. That's an explicit design decision, since:
- the cost is very small (32 bytes for header + map overhead so < 100 bytes)
- it's an infrequent event
- it saves on the complexity of having to recalculate the cache on re-orgs (eg compare with https://github.com/bitcoin-core-review-club/bitcoin/commit/af3bddfbdfb7de06d6e63a0a425da483c435521a#diff-eff7adeaec73a769788bb78858815c91R1258)

> if CFCHECKPT_INTERVAL would ever change

In that case, we'd need to change the software, which would involve a stop/start. This is an in-memory cache so gets dropped on shutdown.